src/interchange/src/protobuf/decode.rs

```
@@ -233,50 +237,101 @@ fn json_nested_from_serde_value(
+                        }
 
-                            let nested_message_descriptor = match nested_message_descriptor {
-                                FieldType::Message(m) => m,
-                                _ => bail!("Nested message is the wrong type"),
-                            };
+                        let nested_message_descriptor = match nested_message_descriptor {
+                            FieldType::Message(m) => m,
+                            _ => bail!("Nested message is the wrong type"),
+                        };
 
-                            json_nested_from_serde_value(
-                                &v,
-                                packer,
-                                nested_message_descriptor
-                                    .field_by_name(s)
-                                    .expect("Expected this to work"),
-                                descriptors,
-                            )?;
-                        }
-                        _ => bail!("Unrecognized value while trying to parse a nested message"),
+                        json_nested_from_serde_value(
+                            &v,
+                            packer,
+                            nested_message_descriptor
+                                .field_by_name(s)
+                                .expect("nested message to exist"),
+                            descriptors,
+                            typ,
+                        )?;
                     }
+                    _ => bail!("Unrecognized value while trying to parse a nested message"),
                 }
                 Ok(())
-            });
+            };
+
+            match &column_type.scalar_type {
+                ScalarType::Map { value_type, .. } => {
+                    let mut kvs = m.iter().collect::<Vec<_>>();
+                    kvs.sort_by(|(k1, _v1), (k2, _v2)| k1.cmp(k2));
+                    kvs.dedup_by(|(k1, _v1), (k2, _v2)| k1 == k2);
```

I'm a little confused why the dedup is necessary? I would expect that either protobuf records aren't allowed to have dups or that we should throw an error stating that we don't support duplicates if there are dups or something.